### PR TITLE
GE PMT: handle system errors

### PR DESCRIPTION
**Pull request by psadie-sfdo**
# Critical Changes

# Changes

# Issues Closed

# Community Ideas Delivered

# New Metadata
- Custom Labels
  - commonCriticalError
  - commonReviewForm
  - geErrorCardChargedBDIFailed
  - geErrorRequestTimedout
  - geErrorUncertainCardChargePart1
  - geErrorUncertainCardChargePart2
  - geErrorUncertainCardChargePart3
  - geErrorUncertainCardChargePart4
- Lightning Web Components
  - gePurchaseCallModalError
  - utilCustomErrors
# Deleted Metadata

### CODE REVIEW

**Comment by psadie-sfdo** [comment]
**lurch: attach W-039744

**Comment by LurchTheButler** [comment]
Tracking <a href="https://foundation.my.salesforce.com/a2x1E000001HsmjQAC" target="_blank">W-039744</a>

File: src/labels/CustomLabels.labels [hunk 1]
```diff
@@ -5641,6 +5657,14 @@ If preferred, you can disable Gift Entry and instead use older Batch Gift Entry
         <shortDescription>Common error when a total doesn't match Donation Amount.</shortDescription>
         <value>Total amount doesn't match the {0}</value>
     </labels>
+    <labels>
+        <fullName>geErrorCardChargedBDIFailed</fullName>
+        <categories>Gift Entry</categories>
+        <language>en_US</language>
+        <protected>true</protected>
+        <shortDescription>Failure description if card is charged and donation record has not been created</shortDescription>
+        <value>The card has been successfully charged but the payment has not been recorded in Salesforce. Please correct the errors and resubmit the form to record the donation in Salesforce.</value>
```
Comment (melissabarber) [inline src/labels/CustomLabels.labels:5666]: @psadie-sfdo Oh bother. We forgot to run this past Sarah during our call today. @smarovich will you be able to make it to our sprint review tomorrow so I can get a second set of eyes on this?

File: src/labels/CustomLabels.labels [hunk 2]
```diff
@@ -5812,6 +5844,41 @@ If preferred, you can disable Gift Entry and instead use older Batch Gift Entry
...
         <value>Required field</value>
     </labels>
+    <labels>
+        <fullName>geErrorUncertainCardChargePart1</fullName>
+        <categories>Gift Entry</categories>
+        <language>en_US</language>
+        <protected>true</protected>
+        <shortDescription>Failure msg description for system error &amp; uncertain if card charged - Part 1</shortDescription>
+        <!-- {0} = custom label commonPaymentServices i.e. Payment Services -->
+        <value>A system error occurred. Your admin should review transactions in {0} and determine if the card was charged.</value>
```
Comment (melissabarber) [inline src/labels/CustomLabels.labels:5854]:
```suggestion
        <value>A system error occurred for the {0} donation by {1}. Your admin should review transactions in {2} and determine if the card was charged.</value>
```

File: src/labels/CustomLabels.labels [hunk 3]
```diff
@@ -5812,6 +5844,41 @@ If preferred, you can disable Gift Entry and instead use older Batch Gift Entry
         <shortDescription>Field-level error message for required field missing on NPSP Data Import object</shortDescription>
         <value>Required field</value>
     </labels>
+    <labels>
+        <fullName>geErrorUncertainCardChargePart1</fullName>
+        <categories>Gift Entry</categories>
+        <language>en_US</language>
+        <protected>true</protected>
+        <shortDescription>Failure msg description for system error &amp; uncertain if card charged - Part 1</shortDescription>
+        <!-- {0} = custom label commonPaymentServices i.e. Payment Services -->
```
Comment (melissabarber) [inline src/labels/CustomLabels.labels:5853]:
```suggestion
        <!-- {0} = Donation Amount {1} = Donor Name {2} = custom label commonPaymentServices i.e. Payment Services -->
```

File: src/labels/CustomLabels.labels [hunk 4]
```diff
@@ -5812,6 +5844,41 @@ If preferred, you can disable Gift Entry and instead use older Batch Gift Entry
...
+        <value>If the card was not charged, reenter the gift and card payment in Gift Entry.</value>
+    </labels>
+    <labels>
+        <fullName>geErrorUncertainCardChargePart3</fullName>
+        <categories>Gift Entry</categories>
+        <language>en_US</language>
+        <protected>true</protected>
+        <shortDescription>Failure msg description for system error &amp; uncertain if card charged - Part 3</shortDescription>
+        <!-- {0} = custom label commonPaymentServices i.e. Payment Services -->
+        <value>If the card was charged, wait for {0} to send the processed payment to Salesforce, and update the gift and donor records as needed.</value>
```
Comment (melissabarber) [inline src/labels/CustomLabels.labels:5871]:
```suggestion
        <value>If the card was charged, note the Transaction ID in {0}, and search for records with that ID in Salesforce. Update the gift and donor records as needed.</value>
```

File: src/labels/CustomLabels.labels [hunk 5]
```diff
@@ -5812,6 +5844,41 @@ If preferred, you can disable Gift Entry and instead use older Batch Gift Entry
...
+    </labels>
+    <labels>
+        <fullName>geErrorUncertainCardChargePart4</fullName>
+        <categories>Gift Entry</categories>
+        <language>en_US</language>
+        <protected>true</protected>
+        <shortDescription>Failure msg description for system error &amp; uncertain if card charged - Part 4</shortDescription>
+        <!-- {0} = custom label commonPaymentServices i.e. Payment Services -->
+        <value>For more information, please see {0} troubleshooting documentation.</value>
```
Comment (melissabarber) [inline src/labels/CustomLabels.labels:5880]:
```suggestion
        <value>For more information, please see {0} troubleshooting documentation on the Power of Us Hub.</value>
```

**Comment by melissabarber** [comment]
@psadie-sfdo and @ptewson-sfdo I added suggestions to the custom labels based on our many discussions today, and as long as those are implemented, I'm signed off on this PR. Thank you for all your help and patience getting to the bottom of these tricky scenarios!

File: src/pages/GE_TokenizeCard.page [hunk 6]
```diff
@@ -43,9 +43,17 @@
                 '}'
         })
         .then(() => {
+            isLoaded();
+
+            setTimeout(function() {
```
Comment (bdvorachek) [inline src/pages/GE_TokenizeCard.page:48]: Is this a potential fix or workaround for the spinner hang that we see sometimes on first load?

File: src/lwc/geFormWidgetTokenizeCard/geFormWidgetTokenizeCard.js [hunk 7]
```diff
@@ -16,24 +17,50 @@ export default class geFormWidgetTokenizeCard extends LightningElement {
...
-    }
-
     async connectedCallback() {
         this.domain = await getOrgDomain();
         this.visualforceOrigin = this.buildVisualforceOriginUrl(this.domain);
     }
 
     renderedCallback() {
         this.registerPostMessageListener();
+        registerListener('disableGeFormWidgetTokenizeCard', this.handleDisableWidget, this);
```
Comment (bdvorachek) [inline src/lwc/geFormWidgetTokenizeCard/geFormWidgetTokenizeCard.js:33]: Can we move this event name into geConstants.js and import it wherever used? I don't think we need to do this for every event, only ones that are fired from one component and handled in another.

File: src/pages/GE_TokenizeCard.page (hunk 6 above)
Comment (psadie-sfdo) [reply src/pages/GE_TokenizeCard.page:48]: Yeah, a bandaid to the issue until we can look into it more closely :\

File: src/lwc/geFormWidgetTokenizeCard/geFormWidgetTokenizeCard.js (hunk 7 above)
Comment (psadie-sfdo) [reply src/lwc/geFormWidgetTokenizeCard/geFormWidgetTokenizeCard.js]: Updated!

**Comment by psadie-sfdo** [comment]
> @psadie-sfdo and @ptewson-sfdo I added suggestions to the custom labels based on our many discussions today, and as long as those are implemented, I'm signed off on this PR. Thank you for all your help and patience getting to the bottom of these tricky scenarios!

@melissabarber - proposed custom labels are updated along with the necessary logic update.

File: src/classes/GE_PaymentServices.cls [hunk 8]
```diff
@@ -56,6 +56,20 @@ public with sharing class GE_PaymentServices {
 
     @TestVisible Map<String, String> config;
 
+    private enum ElevateEndpoints { TOKENIZE, PURCHASE }
+
+    private String getEndpoint(ElevateEndpoints endpoint) {
```
Comment (vmartinov-sforg) [inline src/classes/GE_PaymentServices.cls:61]: We were using this endpoints from a 'test' class that's why they were public. That class needs to be re-visited anyway, just pointing out that we might consider making the enum public at some point in the future.

File: src/lwc/geFormRenderer/geFormRenderer.js [hunk 9]
```diff
@@ -305,29 +309,75 @@ export default class GeFormRenderer extends NavigationMixin(LightningElement) {
...
-            return;
+            return this.handleCatchOnSave(error);
         }
 
         // Handle any other error
         handleError(error);
     }
 
+    catchHttpRequestError(error) {
+        this.hasPurchaseCallTimedout = error.statusCode === 408;
```
Comment (vmartinov-sforg) [inline src/lwc/geFormRenderer/geFormRenderer.js:334]: Could we use a constant for 408? I believe we are using constants for 201 and 400.

File: src/lwc/gePurchaseCallModalError/gePurchaseCallModalError.js-meta.xml [hunk 10]
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<LightningComponentBundle xmlns="http://soap.sforce.com/2006/04/metadata">
+    <apiVersion>47.0</apiVersion>
```
Comment (vmartinov-sforg) [inline src/lwc/gePurchaseCallModalError/gePurchaseCallModalError.js-meta.xml:3]: Are we at api 48 now?

File: src/lwc/utilCustomErrors/utilCustomErrors.js-meta.xml [hunk 11]
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<LightningComponentBundle xmlns="http://soap.sforce.com/2006/04/metadata">
+    <apiVersion>47.0</apiVersion>
```
Comment (vmartinov-sforg) [inline src/lwc/utilCustomErrors/utilCustomErrors.js-meta.xml:3]:
```suggestion
    <apiVersion>48.0</apiVersion>
```

File: src/classes/GE_PaymentServices.cls (hunk 8 above)
Comment (psadie-sfdo) [reply src/classes/GE_PaymentServices.cls:61]: Ah! Thank you for the reminder. I'll revert these to public for now.

**Pull request merged**